internal/dataplane/sendconfig/dbmode.go

```diff
@@ -267,3 +271,52 @@ func (s *UpdateStrategyDBMode) targetState(
 
 	return state.Get(rawState)
 }
+
+// refillPluginIDs keeps the plugin ID in the target state if there are already the same plugin
+// (identified by plugin name and attached service, route, consumer, consumer group) in the current state.
+// This prevents conflicts during the upgrade where the existing plugins have different IDs with the ID generated in building kong state.
+func (s *UpdateStrategyDBMode) refillPluginIDs(currentState *state.KongState, targetState *state.KongState) error {
+	plugins, err := currentState.Plugins.GetAll()
+	if err != nil {
+		return fmt.Errorf("failed getting plugins in current state for %s: %w", s.client.BaseRootURL(), err)
+	}
+	// For each existing plugin in the DB, we look for the same plugin in the target state and re-fill the ID.
+	for _, existingPlugin := range plugins {
+		var serviceID, routeID, consumerID, consumerGroupID string
+		if existingPlugin.Service != nil && existingPlugin.Service.ID != nil {
+			serviceID = *existingPlugin.Service.ID
+		}
+		if existingPlugin.Route != nil && existingPlugin.Route.ID != nil {
+			routeID = *existingPlugin.Route.ID
+		}
+		if existingPlugin.Consumer != nil && existingPlugin.Consumer.ID != nil {
+			consumerID = *existingPlugin.Consumer.ID
+		}
+		if existingPlugin.ConsumerGroup != nil && existingPlugin.ConsumerGroup.ID != nil {
+			consumerGroupID = *existingPlugin.ConsumerGroup.ID
+		}
+		// If the same plugin is in the target state and we have filled a different ID with the existing plugin,
+		// we re-fill the ID of the plugin in the target state to keep the ID the same as the existing plugin.
+		targetPlugin, err := targetState.Plugins.GetByProp(*existingPlugin.Name, serviceID, routeID, consumerID, consumerGroupID)
+		if err != nil {
+			if !errors.Is(err, state.ErrNotFound) {
+				s.logger.Error(err, "failed to get plugin with given fields in the target state")
+			}
+			continue
+		}
+		if existingPlugin.ID != nil && targetPlugin.ID != nil && *targetPlugin.ID != *existingPlugin.ID {
+			// The memdb to store the target state uses `id` to identify the plugin.
+			// So we need to delete the plugin with the new ID first then insert the same plugin with the same ID as the existing plugin.
+			err = targetState.Plugins.Delete(*targetPlugin.ID)
```

Comment:
REVIEW: should we directly `return err` when we failed to run the deletion and re-insertion?

Reply:
I believe we should. That would mean we'd retain the current behavior when there's an error, right? Which means we'd silently continue on an error and allow broken configuration to propagate.

Q: does this return an error if the target state doesn't contain a plugin with requested ID? If so I believe that particular error should be silenced, right?

Reply:
> I believe we should. That would mean we'd retain the current behavior when there's an error, right? Which means we'd silently continue on an error and allow broken configuration to propagate.
> 
> Q: does this return an error if the target state doesn't contain a plugin with requested ID? If so I believe that particular error should be silenced, right?

If the target state does not contain the plugin with the ID, an `ErrNotFound` is returned. So this particular error can be ignored and other errors should be returned.